acm-certificate self-provisioning module

lzecca78 opened the pull request:
Hi guys, the purpose of this pr is to change the current behaviour of the https listener that depends on an external acm_certificate manually created : https://github.com/sighupio/fury-kubernetes-aws/blob/master/modules/aws-kubernetes/lb_external.tf#L70-L75, to an acm_certificate created by default passing to it some needed variable.
I am analyzing an `happy path` scenario in which kubernetes is setup in a single route_53 zone, but is a realistic one imho. 
The module acm-certificate is totally self provisioned because it will be validated against dns records that terraform will provide, so can be fully provisioned out of the box.
Is  a first step, and i ask you, to review and suggest scenarios to be implemented as well, if you see something missing or wrong.
Thanks guys for your patience 🙏 !

angelbarrera92 commented on modules/aws-kubernetes/acm-certificate/main.tf:
```diff
@@ -0,0 +1,49 @@
+}
+
+## acm certificate records for validation
+resource "aws_route53_record" "main-0" {
```
Why do you create `main-0` and `main-1` aws_rourte53_record? Can it be achieve with a `count` attribute and a single resource `main`?

lzecca78 replied:
yep, you're right! this is a stylish refactoring (i will change that asap) even first as i asked, i would like what do you think about the idea of changing the behaviour mainly

lzecca78 replied:
i am always of the idea of 

1. make it works
1. make it fast
1. make it better

now the pr is even before the point 1 afaik, that is "could be an acceptable implementation?"

angelbarrera92 replied:
Sure! You're right.

For sure, all movements to be more "automated" it's welcomed so changing the current behavior (having a certificate already available in acm) to one more automated (generating a certificate) it's nice.

Consider to make "migration" tests (after making it working). Thanks for your work!

ralgozino commented:
Hello Luca,

I think that as a first attemp it is quite an improvement, I would like to have the "happy path" you are assuming written somewhere, either as comments on the Terraform code or in the README file maybe.

lzecca78 commented:
@ralgozino, as the previous acm implementation (where the assumption was: "i want that the user identify a manual acm certificate and pass it to my config", and there weren't any comments in the code, i understand it reading the code), i started make some assumptions as well:

1. a single kubernetes cluster `owns` a domain or multiple domains
2. acm doesn't exist at the moment of the cluster setup
3. being true the 1st assumption, acm will be creating  passing as variables the list of domains

I am agree with you that should be better to have comment for each terraform resource in order to understand easily how that resource behaves. but i guess that this should be an improvement that deserves a dedicated pr

lzecca78 commented:
@lnovara thanks for reply first of al. I am agree with your reply, one of my first doubt as exposed in the pr was if the pr makes some sense in the fury context. As i said today, don't know anything of which is the direction taken by you for the fury product, don't knowing anything at all, i just open this, starting my assumption from the code i read, that was my only possible starting point at my current situation. Btw in a way or another, if you want a public dns zone, both scenarios are possible as well, the acm-certificate and the cert-manager one, even if the cert-manager solution is a cloud-agnostic one and surely can be re-used everywhere (i never used it so far).  Depending on the whishes of the customer (that could enforce the acm certficate because doesn't like letsencrypt) could be merged this solution by configuring the resource creation depending on a boolean variable passed by module that by default could be false. Anyway, thank for making me more clear the scenario, in respect of which if i was aware before, probably i would have developed in a total different way.  Develop something blindly, often ends in a good code probably, but poor solution

lzecca78 commented:
Hi guys, hello here again. For Camelot customer, we are going to need several acm certificate (as @ralgozino  know) and me and @nutellinoit will need something to automate the process of this task.
I refactored all this PR in order to make it the less invasive as possible, adding it as non mandatory, and leaving the previous logic untouched. Basically this is the proposal (as written in the commit):

```  Proposal: due to the need of have acm generated and validated, i added
    the following logic with the following constraints:
    1) you have a public_zone_id for the acm domain you want to generate
    2) the data "aws_acm_certificate" will wait for its generation
    3) the resource is not mandatory, unless you don't define at least one
    element of the variable "subject_alternative_names" the resource won't be
    executed at all
    4) the resource, once it has been created, will be used by the data
    "aws_acm_certificate" (for this reason i added the state PENDING_VALIDATION as well )
    4) i would like that someone help me for the testing workflow
```
For sure this is far from being a perfect solution or the silver bullet one, but, really, is important for us to be able to automate this process, and probably this will be useful not only for this costumer imho.
Please 🙏 review again this modification and talk about that. We need it asap because on 15th January will need to create them. Looking forward for your feedback guys! 🙏

angelbarrera92 commented:
LGTM!

lzecca78 commented:
i am agree. something in a repo like `fury-kubernetes-aws-acm-certificate` ?

angelbarrera92 commented:
> i am agree. something in a repo like `fury-kubernetes-aws-acm-certificate` ?

I think, @lnovara suggested to put it into this repository but in a separate terraform module. Am i right @lnovara?

lnovara commented:
You're right @angelbarrera92.

@lzecca78 I was suggesting of moving this module from `fury-kubernetes-aws/modules/aws-kubernetes/acm-certificate` to `fury-kubernetes-aws/modules/acm-certificate` in order for the module to be a first-class citizen and not "hidden" inside another module as it is useful per se and not strictly bound to `aws-kubernets`.

lzecca78 commented:
as requested in the latest commit i moved the module in `modules/` with an example_usage.tf commented.

lnovara commented:
@lzecca78 Instead of having a fully commented `example_usage.tf`, why not add a `README.md` with some documentation which could be based off the content of `example_usage.tf`?

ralgozino commented:
One question that came to my mind.. shouldn't we use cert-manager? Is it not applicable to this case?

lzecca78 commented:
@lnovara i converted the tf example into the README.md,  now seems better i guess.
@ralgozino , probably yes for some cases cert-manager could be a solution, but for aws setup i guess that acm will be by far the simplest/ 0 effort and 0 cost solution. Having a module that make life simpler in those cases, i guess that could be a great added value

angelbarrera92 commented:
INHO, we have to think about how we manage certificates. We have to differentiate "required" certificates (required to setup the cluster) vs "optional" certificates.

If we can avoid generate certificates apart `cert-manager`, we should do it. 

**But again, all I said has to be discussed and analyzed carefully.**

At this point, the PR LGTM! but... i don't know if at some point in time it will be unnecessary.

lzecca78 commented:
@angelbarrera92  i understand perfectly your point, but, is not totally our decision, but customer's as well. I mean, if a customer wants to manage certificate with ACM of AWS, we must be able to handle them and handle them easily. This are my 2 cents :)

lnovara commented:
I agree with @lzecca78, having a Terraform module to manage ACM certificates could come in handy.

Do we all agree that this PR can merged?

lzecca78 commented:
just to push this pr, can we merge?

angelbarrera92 commented:
> just to push this pr, can we merge?

I think so ;)

lnovara commented:
Shouldn't the target of this PR be `fury-kubernetes-aws-extra` (or something like that)?

For more context, see #30 and #21.

angelbarrera92 commented:
> Shouldn't the target of this PR be `fury-kubernetes-aws-extra` (or something like that)?
> 
> For more context, see #30 and #21.

Uhmmmm. Yes, we are refactoring this kind of modules. So, hold-on until we create the right repository

lzecca78 commented:
Yeah @lnovara , i am agree, but when this PR born, the discussions you are referring to, didn't still exist. So let's wait for a decision and then we'll put into it.

lzecca78 commented:
hi @lnovara, @ralgozino, @angelbarrera92, does this oooooooold PR is still relevant?

lnovara commented:
Not relevant anymore. Closing.

For anyone looking for a Terraform module to manage ACM certificate: https://github.com/terraform-aws-modules/terraform-aws-acm
